File: issue-tracker-web/src/types/notification.ts
export enum NotificationType {
  PROJECT_CREATED = 'PROJECT_CREATED',
  PROJECT_DELETED = 'PROJECT_DELETED',
  TICKET_CREATED = 'TICKET_CREATED',
  TICKET_STATUS_CHANGED = 'TICKET_STATUS_CHANGED',
  TICKET_ASSIGNED = 'TICKET_ASSIGNED',
  TICKET_DELETED = 'TICKET_DELETED'
}

export interface Notification {
  id: number;
  type: NotificationType;
  title: string;
  message: string;
  userId: number | null;
  departmentId: number | null;
  departmentName: string | null;
  relatedEntityId: number | null;
  isRead: boolean;
  createdAt: string;
}

export interface NotificationResponse {
  count: number;
}

File: src/main/java/com/suryakn/IssueTracker/entity/Notification.java
package com.suryakn.IssueTracker.entity;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDateTime;

@Entity
@EntityListeners(AuditingEntityListener.class)
@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
public class Notification {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Enumerated(EnumType.STRING)
    private NotificationType type;

    private String message;

    private String title;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    @JsonIgnoreProperties({"password", "role", "comments", "createdTickets", "assignedTickets"})
    private UserEntity user;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "department_id")
    @JsonIgnoreProperties({"projects", "users"})
    private Department department;

    private Long relatedEntityId; // ID of the related ticket/project

    @Builder.Default
    private Boolean isRead = false;

    @CreatedDate
    private LocalDateTime createdAt;

    public Notification(NotificationType type, String title, String message, UserEntity user, Department department, Long relatedEntityId) {
        this.type = type;
        this.title = title;
        this.message = message;
        this.user = user;
        this.department = department;
        this.relatedEntityId = relatedEntityId;
        this.isRead = false;
    }
}

